### DISCUSSION TITLE
client.dns.lookup 설정은 어디에 할까?

### DISCUSSION THREAD

**Discussion by kyungjunleeme** [kyungjunleeme/kafka_study#37] [Q&A] [answered]
5장 관련 내용

**Comment by kyungjunleeme** [comment]
제가 찾은 바에 따르면 client 쪽(producer, consumer, adminclient 쪽에서만 하면 될것 같고 필요할때)
msk 와 같은 환경에서는 보통 lb 가 앞에 붙은 상황이니, client 쪽에도 이를 고려해주면 더 재해 복구가 및 가용성 정도가 더 커질듯

관련
KIP
https://cwiki.apache.org/confluence/display/KAFKA/KIP-302+-+Enable+Kafka+clients+to+use+all+DNS+resolved+IP+addresses
https://cwiki.apache.org/confluence/display/KAFKA/KIP-602%3A+Change+default+value+for+client.dns.lookup
https://cwiki.apache.org/confluence/display/KAFKA/KIP-235%3A+Add+DNS+alias+support+for+secured+connection

우선 기본값은 - `use_all_dns_ips`

https://github.com/kyungjunleeme/kafka_study/blob/main/KAFKA/chapter05/%EA%B2%BD%EC%A4%80.md

경준 설명 자료 추가보충

**Comment by JONGWOONG-SEON** [comment]
좋은 자료 공유 감사합니다. 자료를 확인하니 클라이언트 부분에서만 설정해주면 되는 것 같습니다.

해당 자료를 보니까 129페이지 내용이 조금 더 이해되는거 같아 부족하지만 의견 남기겠습니다.

개요에 명시되어 있던 부분을 가볍게 넘겼던게 화근인 것 같습니다.

1. DNS 별칭을 사용하는 부분 - 개별 브로커가 별칭으로 구성된 경우
2. 다수의 IP 주소로 연결되는 DNS 이름을 사용하는 경우 - 2개 이상의 IP 주소로 연결되는 하나의 DNS 항목을 사용할 경우

`default`설정에 대한 부분을 짚고 넘어가야 할 것 같은데, bootstrap.servers 설정에서 broker 주소를 명시하여 client 측에서 broker 측 특정 클러스터나 전체 클러스터에 주소를 알게 하여, 순회 시 브로커와 최초 연결 후 작업을 종료하는 것 같습니다.
broker1.hostname.com, broker2.hostname.com ...와 같이 작성 할 수도 있으나
편의 상 *.hostname.com 와 같이 대체하여 사용하는 상황에서

SASL 인증 설정 시 , *.hostname.com 으로 명시하면 외부 공격으로 인지하여 인증을 거부하는 일이 발생하나 봅니다.

그래서 추가된 옵션이 `client.dns.lookup=resolve_canonical_bootstrap_server_only` 인 것 같습니다.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-235%3A+Add+DNS+alias+support+for+secured+connection
를 살펴보면 
```
이슈
bootstrap.server에서 dns 별칭을 지정할 때 Java 클라이언트 API는 그 뒤에 있는 모든 CNAME을 확인하지 않습니다.

수정사항
클라이언트 코드 변경 :
ClientUtils에서 parseAndValidateAddresses()를 변경하여 전체 DNS 확인을 허용하면 모든 기본 호스트가 Kafka 노드로 추가됩니다. 
이를 통해 bootstrap.servers에서 별칭을 사용할 수 있습니다. 
기존 사용자에게 이 동작을 강제로 적용하는 것은 SSL 인증을 손상시킬 수 있으므로 바람직하지 않습니다. 
따라서 이는 선택 사항이어야 합니다.

관련 옵션
Implemented with a ClientDnsLookup enum including values : 
RESOLVE_CANONICAL_BOOTSTRAP_SERVERS_ONLY("resolve_canonical_bootstrap_servers_only")
```

라고 되어있는 걸로 보아 *.hostname.com 을 기본 허용 주체로 반환하는 것 같습니다.
DNS resolver 에 이념을 일부 상속하는 것 같아 하기에 이미지로 첨부합니다.
![image](https://github.com/user-attachments/assets/3ef5f865-09f5-4049-a662-16b673a3739d)

즉, `resolve_canonical_bootstrap_server_only` 설정은 분산된 호스트명을 도메인에 CNAME 매핑을 부여하여 연결하는 방식,
`use_all_dns_ips` 는 DNS 기반 로드밸런스 시 DNS 주소와 매핑되어 있는 다중 ip를 반환하여 연결하는 방식.
으로 이해했습니다.

클라이언트 측에서만 설정해도 되는 옵션은 맞는 거 같습니다

**Reply by kyungjunleeme** [reply]
@JONGWOONG-SEON 오오!!! 종웅님 확인 감사합니다.

**Comment by kyungjunleeme** [comment]
@hellices 멘토님 혹시 해당 내용 한번 확인 부탁드려도 괜찮을까요?

**Reply by hellices** [reply, marked as answer]
앗, 편하게 태깅하셔도 됩니다.
[client.dns.lookup](https://kafka.apache.org/documentation/#adminclientconfigs_client.dns.lookup)을 kafka 홈페이지에서 조회해보면 producer, consumer, adminclient, connect 등에서 설정하는 것으로 나옵니다.

저희쪽 설정도 한번 확인해봤습니다.
strimzi에서는 특별히 client.dns.lookup에 대한 설정을 가이드하지 않고 있습니다.
strimzi로 kubernetes에 설치를 하면 3대의 broker와 총 네 개의 외부 도메인이 생성됩니다.
- external bootstrap domain
- external domain1(broker1)
- external domain2(broker2)
- external domain3(broker3)

연동하는 대상 시스템에는 external bootstrap domain을 전달해줍니다. 
해당 도메인을 확인해보니 ingress -> service -> pod(broker1), pod(broker2), pod(broker3) 이렇게 세 개가 물려서 k8s service에서 로드밸런싱을 하는 것으로 나옵니다.
kubernetes 내부용 internal bootstrap service에도 동일하게 pod(broker1), pod(broker2), pod(broker3)이 모두 연결되어 있어서 자연스럽게 로드밸런싱이 될 것 같습니다.

즉, strimzi에서는 bootstrap 서비스는 하나의 ip만 가지고 있고, kubernetes network에서 로드밸런싱을 자연스럽게 할 것으로 예상됩니다.

**Reply by kyungjunleeme** [reply]
답변 감사합니다. 저도 종웅님이랑 인환님이 알려주신 내용을 참고했을때, 이 설정의 핵심은 설정을 하는것이 중요한게아니라 이미 기본값설정은되어 있는데, 내부에서 어떻게 자연스럽게 구성요소간의 통신이 되면서 자연스럽게 로드밸런싱이 되냐와 관련된 핵심 설정 정도로 이해하면 좋겠네요.

**Reply by kyungjunleeme** [reply]
25.03.28 - 내용 보충
https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java#L795

```
질문의 시나리오와 코드 일치 여부: 제공된 코드에 따르면, external bootstrap domain은 Bootstrap Service로 모든 브로커에 로드밸런싱되고, external domain1, domain2, domain3은 브로커별 Service로 각각 매핑됩니다. 이는 질문과 일치합니다.
Service 개수: 코드상 외부 리스너(ingress) 기준으로 4개 Service 생성:
my-cluster-kafka-bootstrap (Bootstrap).
my-cluster-kafka-0 (broker1).
my-cluster-kafka-1 (broker2).
my-cluster-kafka-2 (broker3).
내부용: 이 코드 조각에는 없지만, type: internal 리스너가 있다면 추가 Service(또는 포트 구분)로 처리됨.
코드와 매핑
external bootstrap domain: generateExternalBootstrapServices() -> my-cluster-kafka-bootstrap.
external domain1, domain2, domain3: generatePerPodServices() -> my-cluster-kafka-0, -1, -2.
```

**Comment by kyungjunleeme** [comment]
25.03.28 - 내용 보충
https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java#L795
